numpy/doc/subclassing.py

```diff
@@ -412,15 +414,160 @@ def __array_finalize__(self, obj):
   >>> v.info
   'information'
 
-.. _array-wrap:
+.. _array-ufunc:
+
+``__array_ufunc__`` for ufuncs
+------------------------------
+
+  .. versionadded:: 1.13
+
+A subclass can override what happens when executing numpy ufuncs on it by
+overriding the default ``ndarray.__array_ufunc__`` method. This method is
+executed *instead* of the ufunc and should return either the result of the
+operation, or :obj:`NotImplemented` if the operation requested is not
+implemented.
+
+The signature of ``__array_ufunc__`` is::
+
+    def __array_ufunc__(ufunc, method, *inputs, **kwargs):
 
-``__array_wrap__`` for ufuncs
--------------------------------------------------------
+    - *ufunc* is the ufunc object that was called.
+    - *method* is a string indicating how the Ufunc was called, either
+      ``"__call__"`` to indicate it was called directly, or one of its
+      :ref:`methods<ufuncs.methods>`: ``"reduce"``, ``"accumulate"``,
+      ``"reduceat"``, ``"outer"``, or ``"at"``.
+    - *inputs* is a tuple of the input arguments to the ``ufunc``
+    - *kwargs* contains any optional or keyword arguments passed to the
+      function. This includes any ``out`` arguments, which are always
+      contained in a tuple.
+
+A typical implementation would convert any inputs or ouputs that are
+instances of one's own class, pass everything on to a superclass using
+``super()``, and finally return the results after possible
+back-conversion. An example, taken from the test case
+``test_ufunc_override_with_super`` in ``core/tests/test_umath.py``, is the
+following.
+
+.. testcode::
+
+    input numpy as np
+
+    class A(np.ndarray):
+        def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+            args = []
+            in_no = []
+            for i, input_ in enumerate(inputs):
+                if isinstance(input_, A):
+                    in_no.append(i)
+                    args.append(input_.view(np.ndarray))
+                else:
+                    args.append(input_)
+
+            outputs = kwargs.pop('out', None)
+            out_no = []
+            if outputs is not None:
+                out_args = []
+                for j, output in enumerate(outputs):
+                    if isinstance(output, A):
+                        out_no.append(j)
+                        out_args.append(output.view(np.ndarray))
+                    else:
+                        out_args.append(output)
+                kwargs['out'] = tuple(out_args)
+
+            info = {}
+            if in_no:
+                info['inputs'] = in_no
+            if out_no:
+                info['outputs'] = out_no
+
+            results = super(A, self).__array_ufunc__(ufunc, method,
```

Comment:
Obviously the 1-tuple should still be unpacked at the top level for calls like `np.add(a, b)`, but I think it's a lot easier to work with `array_ufunc` if the model is:
* You get a tuple of inputs (`*args`, `len(args) == ufunc.nin`)
* You get a tuple of preallocated outputs (`out`, `len(out) == ufunc.nout`)
* You are expected to produce a tuple of final outputs (`results`, `len(results)  == ufunc.nout`)

There are two groups of people this change would affect:
* People who care about multi-output ufuncs &mdash; these people no longer have to special-case multi-result ufuncs
* People who don't care about multi-output ufuncs &mdash; these people have to write a little more code, but are prevented from writing code that crashes on multi-result ufuncs

This change makes it easier to do complex things, and harder to do incorrect things - which sounds like a win to me